### ISSUE TITLE
FP field_reassign_with_default with macros

### ISSUE THREAD

**Issue by xu-cheng** [rust-lang/rust-clippy#7155]
Lint name: field_reassign_with_default

I tried this code (https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=fb08ca7dbb9fb018ef6927cf078ed00b):

```rust
use std::collections::HashSet;

#[derive(Debug, Default)]
struct MyData {
    field1: Option<i32>,
    field2: HashSet<i32>,
}

macro_rules! create_my_data {
    ($($key:ident: $value:tt,)+) => { create_my_data!($($key: $value),+) };
    ($($key:ident: $value:tt),*) => {
        {
            let mut data = $crate::MyData::default();
            $(
                create_my_data!(@data data @parse $key $value);
            )*
            data
        }
    };
    (@data $data:ident @parse field1 $x:expr) => {
        $data.field1 = Some($x);
    };
    (@data $data:ident @parse field2 [ $($v:expr,)+ ]) => {
        create_my_data!(@data $data @parse field2 [ $($v),+ ]);
    };
    (@data $data:ident @parse field2 [ $($v:expr),* ]) => {
        {
            $(
                $data.field2.insert($v);
            )*
        }
    };
}

fn main() {
    let data1 = create_my_data! {
        field2: [1, 2, 3],
    };
    println!("{:?}", data1);
    let data2 = create_my_data! {
        field1: 1,
        field2: [1, 2, 3]
    };
    println!("{:?}", data2);
}
```

I expected to see this happen: clippy reports no issue.

Instead, this happened: 
```
  Checking playground v0.0.1 (/playground)
warning: field assignment outside of initializer for an instance created with Default::default()
  --> src/main.rs:21:9
   |
21 |           $data.field1 = Some($x);
   |           ^^^^^^^^^^^^^^^^^^^^^^^^
...
40 |       let data2 = create_my_data! {
   |  _________________-
41 | |         field1: 1,
42 | |         field2: [1, 2, 3]
43 | |     };
   | |_____- in this macro invocation
   |
   = note: `#[warn(clippy::field_reassign_with_default)]` on by default
note: consider initializing the variable with `MyData { field1: create_my_data! {
        field1: 1,
        field2: [1, 2, 3]
    }, ..Default::default() }` and removing relevant reassignments
  --> src/main.rs:13:13
   |
13 |               let mut data = $crate::MyData::default();
   |               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
...
40 |       let data2 = create_my_data! {
   |  _________________-
41 | |         field1: 1,
42 | |         field2: [1, 2, 3]
43 | |     };
   | |_____- in this macro invocation
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#field_reassign_with_default
   = note: this warning originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

warning: 1 warning emitted

    Finished dev [unoptimized + debuginfo] target(s) in 0.61s
```

### Meta

- `cargo clippy -V`:  clippy 0.1.53 (4de7572 2021-05-01)
